### ISSUE TITLE
Implement the blobParts argument to the Blob constructor correctly.

### ISSUE THREAD

**Issue by Ms2ger** [servo/servo#8397]
Depends on #544.
Code: `components/script/dom/webidls/Blob.webidl`, `components/script/dom/blob.rs`.
Spec: https://w3c.github.io/FileAPI/#constructorBlob

**Comment by frewsxcv** [comment]
#544 is now done

**Comment by jdm** [comment]
This should cover turning the existing `DOMString blobParts` argument into the more correct `sequence<(Blob or DOMString)>` instead, and processing it according to the specification.

**Comment by stspyder** [comment]
Is someone working on this? If no, can I pick this up?

**Comment by jdm** [comment]
Looks like #9360 has already been submitted for this. Sorry!

**Comment by KiChjang** [comment]
@jdm #9360 doesn't seem to be touching anything about the Blob, did you mean to link to another issue instead?

**Comment by jdm** [comment]
Oops! Go for it, @stspyder!

**Comment by stspyder** [comment]
Great!!

**Comment by stspyder** [comment]
I did implement the new constructor with a `Vec<BlobOrString>` but [this wpt test](https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/FileAPI/blob/Blob-constructor.html#L42) times out. I think the test crashes, but is being timed out by the test harness. Any help figuring this out would be appreciated!!
[blob_blobparts_constructor.txt](https://github.com/servo/servo/files/105079/blob_blobparts_constructor.txt)

**Comment by jdm** [comment]
@stspyder Have you tried narrowing down which case triggers the timeout?

**Comment by stspyder** [comment]
@jdm "Passing non-objects, Dates and RegExps for blobParts should throw a TypeError." is the test that causes the timeout. It doesn't get past passing null to Blob constructor. I tried debugging and had unexpected crashes with exit code of 8 when in the `_constructor` of BlobBinding.rs. But sometimes I was able to see that the conversion of the 1st argument fails and the method returns false.

**Comment by jdm** [comment]
```
14:40 <jdm> stspyder: oho - https://gist.github.com/stspyder/19e49baf536cf0517938#file-blobbinding-rs-L23 and https://gist.github.com/stspyder/19e49baf536cf0517938#file-blobbinding-rs-L30 would cause these problems
14:40 <jdm> stspyder: since they're missing https://gist.github.com/stspyder/19e49baf536cf0517938#file-blobbinding-rs-L37
14:42 <jdm> stspyder: well, actually they're missing a throw_type_error call
```

**Comment by nox** [comment]
@jdm No, that `throw_dom_exception` is for the actual return value of the method. Instead we should call `throw_type_exception` from the implementation of `FromJSValConvertible` on `Vec<T>`, like it is done in the inner `BlobOrString` impl.

**Comment by nox** [comment]
Oh I guess you mentioned the throw_type_error part.

**Comment by Ms2ger** [comment]
https://github.com/servo/rust-mozjs/issues/237

**Comment by nox** [comment]
That PR landed a while ago.

**Comment by stspyder** [comment]
Well, sorry it took me a while to come back on this. But most of my work is complete already. I'll submit a PR shortly. Don't want someone to duplicate efforts here.

**Comment by jdm** [comment]
@stspyder See also #9977 that has already been opened :/

**Comment by stspyder** [comment]
@jdm, yes noticed it. #9977 requires changes in tests for the wpt tests to pass which are available in #9979. I'll close #9979 if those changes can be done in #9977.

**Comment by jdm** [comment]
Fixed by #9979.
